File: android/app/src/main/kotlin/com/aispotlight/android/data/SpendLedger.kt
package com.aispotlight.android.data

import android.content.Context
import android.content.SharedPreferences
import androidx.room.Dao
import androidx.room.Entity
import androidx.room.Index
import androidx.room.Insert
import androidx.room.PrimaryKey
import androidx.room.Query
import com.aispotlight.android.R
import com.aispotlight.android.core.Diagnostics
import com.aispotlight.android.core.TokenUsage
import com.aispotlight.android.providers.PricingCatalog
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import java.util.Calendar
import java.util.Locale
import java.util.UUID

/**
 * Append-only cost ledger — the Android port of the macOS `SpendLedger`.
 * Lives in the main Room database (own table; explicit 2→3 migration so
 * existing chats survive the schema bump). Deleting a conversation never
 * touches spend history.
 */

/** What a spend record paid for (mirrors the macOS `SpendKind`). */
enum class SpendKind(val raw: String) {
    CHAT("chat"),       // a model turn (incl. agentic tool loop)
    SUMMARY("summary"), // context-compression summarization call
    OCR("ocr"),         // Mistral OCR (per page)
    STT("stt"),         // speech-to-text (per minute)
    SEARCH("search"),   // Brave web search (per query)
    IMAGE("image");     // ImageAddon cloud operation (fal.ai)

    companion object {
        fun fromRaw(raw: String): SpendKind = entries.firstOrNull { it.raw == raw } ?: CHAT
    }
}

@Entity(tableName = "spend_records", indices = [Index("timestamp")])
data class SpendRecordEntity(
    @PrimaryKey val id: String,
    val timestamp: Long,
    /** SpendKind.raw */
    val kind: String,
    val provider: String,
    val model: String,
    val inputTokens: Int,
    val outputTokens: Int,
    val cacheReadTokens: Int,
    val cacheWriteTokens: Int,
    val reasoningTokens: Int,
    /** Non-token quantity: OCR pages, STT minutes, search queries, images. */
    val units: Double,
    /** null = tokens recorded but no price known for the model at write time. */
    val costUSD: Double?,
    /** true when the stream ended without provider usage and tokens are estimated. */
    val isEstimate: Boolean,
) {
    val usage: TokenUsage
        get() = TokenUsage(inputTokens, outputTokens, cacheReadTokens, cacheWriteTokens, reasoningTokens)
}

@Dao
interface SpendDao {
    @Insert
    suspend fun insert(record: SpendRecordEntity)

    @Query("SELECT * FROM spend_records WHERE timestamp >= :from AND timestamp < :to ORDER BY timestamp")
    suspend fun recordsBetween(from: Long, to: Long): List<SpendRecordEntity>

    @Query("SELECT MIN(timestamp) FROM spend_records")
    suspend fun earliestTimestamp(): Long?

    @Query("SELECT COALESCE(SUM(costUSD), 0) FROM spend_records WHERE timestamp >= :from AND timestamp < :to")
    suspend fun costBetween(from: Long, to: Long): Double
}

/**
 * Recording façade + live counters for the Costs screen (port of the macOS
 * `SpendStore`). Fire-and-forget [record] appends to the ledger, keeps the
 * session/month counters fresh, kicks the weekly price refresh, and returns
 * a budget-warning message when a threshold was crossed (80% / 100%, once
 * per month each — soft limit, never blocks).
 */
object SpendTracker {
    private lateinit var appContext: Context
    private lateinit var prefs: SharedPreferences
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _sessionUSD = MutableStateFlow(0.0)
    val sessionUSD: StateFlow<Double> = _sessionUSD
    private val _monthUSD = MutableStateFlow(0.0)
    val monthUSD: StateFlow<Double> = _monthUSD
    private val _todayUSD = MutableStateFlow(0.0)
    val todayUSD: StateFlow<Double> = _todayUSD

    private val _monthlyBudgetUSD = MutableStateFlow(0.0)
    val monthlyBudgetUSD: StateFlow<Double> = _monthlyBudgetUSD

    private fun dao(): SpendDao = AppDatabase.get(appContext).spendDao()

    /** Called once from App.onCreate. */
    fun init(context: Context) {
        appContext = context.applicationContext
        prefs = appContext.getSharedPreferences("spend", Context.MODE_PRIVATE)
        _monthlyBudgetUSD.value = prefs.getFloat("monthlyBudgetUSD", 0f).toDouble()
        refreshTotals()
    }

    fun setMonthlyBudget(value: Double) {
        _monthlyBudgetUSD.value = value.coerceAtLeast(0.0)
        prefs.edit().putFloat("monthlyBudgetUSD", _monthlyBudgetUSD.value.toFloat()).apply()
    }

    /** Reloads the month/today counters from the ledger (call on screen open). */
    fun refreshTotals() {
        scope.launch {
            val (monthStart, monthEnd) = monthBounds(System.currentTimeMillis())
            _monthUSD.value = dao().costBetween(monthStart, monthEnd)
            _todayUSD.value = dao().costBetween(dayStart(System.currentTimeMillis()), monthEnd)
        }
    }

    /**
     * Appends one record. Safe from any thread; the insert is async, the
     * counter/budget bookkeeping is immediate. Returns the localized budget
     * warning to surface in the chat, or null.
     */
    fun record(
        kind: SpendKind,
        provider: String,
        model: String,
        usage: TokenUsage = TokenUsage(),
        units: Double = 0.0,
        costUSD: Double?,
        isEstimate: Boolean = false,
    ): String? {
        val entity = SpendRecordEntity(
            id = UUID.randomUUID().toString(),
            timestamp = System.currentTimeMillis(),
            kind = kind.raw,
            provider = provider,
            model = model,
            inputTokens = usage.inputTokens,
            outputTokens = usage.outputTokens,
            cacheReadTokens = usage.cacheReadTokens,
            cacheWriteTokens = usage.cacheWriteTokens,
            reasoningTokens = usage.reasoningTokens,
            units = units,
            costUSD = costUSD,
            isEstimate = isEstimate,
        )
        scope.launch { dao().insert(entity) }
        PricingCatalog.refreshIfStale()
        Diagnostics.log(
            "spend",
            "append kind=${kind.raw} provider=$provider model=$model " +
                "cost=${costUSD?.let { String.format(Locale.US, "%.6f", it) } ?: "nil"} est=$isEstimate"
        )

        val cost = costUSD ?: 0.0
        _sessionUSD.value += cost
        _todayUSD.value += cost
        val before = _monthUSD.value
        _monthUSD.value = before + cost
        return budgetWarning(before, _monthUSD.value)
    }

    /** One warning per threshold per month: 80% and 100% of the budget. */
    private fun budgetWarning(before: Double, after: Double): String? {
        val budget = _monthlyBudgetUSD.value
        if (budget <= 0) return null
        val monthKey = monthKey(System.currentTimeMillis())
        for ((fraction, suffix, res) in listOf(
            Triple(1.0, "100", R.string.costs_budget_hit),
            Triple(0.8, "80", R.string.costs_budget_near),
        )) {
            val threshold = budget * fraction
            if (after < threshold || before >= threshold) continue
            val flag = "warned$suffix.$monthKey"
            if (prefs.getBoolean(flag, false)) return null
            prefs.edit().putBoolean(flag, true).apply()
            return appContext.getString(res, usd(after), usd(budget))
        }
        return null
    }

    // MARK: Helpers

    fun usd(value: Double): String =
        if (value < 0.01 && value > 0) String.format(Locale.US, "$%.4f", value)
        else String.format(Locale.US, "$%.2f", value)

    fun monthBounds(at: Long): Pair<Long, Long> {
        val cal = Calendar.getInstance().apply {
            timeInMillis = at
            set(Calendar.DAY_OF_MONTH, 1)
            set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
        }
        val start = cal.timeInMillis
        cal.add(Calendar.MONTH, 1)
        return start to cal.timeInMillis
    }

    private fun dayStart(at: Long): Long = Calendar.getInstance().apply {
        timeInMillis = at
        set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
        set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
    }.timeInMillis

    private fun monthKey(at: Long): String {
        val cal = Calendar.getInstance().apply { timeInMillis = at }
        return "${cal.get(Calendar.YEAR)}-${cal.get(Calendar.MONTH) + 1}"
    }
}

File: Cuate/Models/SpendLedger.swift
import Foundation
import Combine
import SwiftData

// MARK: - Spend records
//
// Append-only cost ledger, deliberately SEPARATE from the chat store: deleting
// a conversation must not erase spend history, and the chat schema stays
// untouched. Same access pattern as ChatPersistence — one serial queue, a
// fresh ModelContext per operation, value types across the boundary.

/// What a spend record paid for.
enum SpendKind: String, CaseIterable {
    case chat      // a model turn in the panel (incl. agentic tool loop)
    case summary   // context-compression summarization call
    case ocr       // Mistral OCR (per page)
    case stt       // speech-to-text (per minute)
    case search    // Brave web search (per query; billed by plan, cost 0)
    case image     // ImageAddon cloud operation (fal.ai)
}

@Model
final class SDSpendRecord {
    @Attribute(.unique) var id: UUID
    var timestamp: Date
    var kindRaw: String
    var provider: String
    var model: String
    var inputTokens: Int
    var outputTokens: Int
    var cacheReadTokens: Int
    var cacheWriteTokens: Int
    var reasoningTokens: Int
    /// Non-token quantity: OCR pages, STT minutes, search queries, images.
    var units: Double
    /// nil = tokens recorded but no price known for the model at write time.
    var costUSD: Double?
    /// true when the stream ended without provider usage (cancel/error) and
    /// the tokens are a script-aware estimate, not an API-reported count.
    var isEstimate: Bool

    init(id: UUID = UUID(), timestamp: Date = Date(), kindRaw: String,
         provider: String, model: String,
         inputTokens: Int = 0, outputTokens: Int = 0, cacheReadTokens: Int = 0,
         cacheWriteTokens: Int = 0, reasoningTokens: Int = 0,
         units: Double = 0, costUSD: Double? = nil, isEstimate: Bool = false) {
        self.id = id
        self.timestamp = timestamp
        self.kindRaw = kindRaw
        self.provider = provider
        self.model = model
        self.inputTokens = inputTokens
        self.outputTokens = outputTokens
        self.cacheReadTokens = cacheReadTokens
        self.cacheWriteTokens = cacheWriteTokens
        self.reasoningTokens = reasoningTokens
        self.units = units
        self.costUSD = costUSD
        self.isEstimate = isEstimate
    }
}

/// Value-type snapshot of a record (crosses the queue boundary; feeds charts).
struct SpendRecordValue: Identifiable {
    let id: UUID
    let timestamp: Date
    let kind: SpendKind
    let provider: String
    let model: String
    let usage: TokenUsage
    let units: Double
    let costUSD: Double?
    let isEstimate: Bool
}

// MARK: - Ledger (persistence)

nonisolated enum SpendLedger {

    static let container: ModelContainer = {
        let schema = Schema([SDSpendRecord.self])
        let url = ChatStore.baseDirectory.appendingPathComponent("CuateSpend.store")
        let config = ModelConfiguration(schema: schema, url: url)
        do {
            return try ModelContainer(for: schema, configurations: config)
        } catch {
            // Same corrupt-store fallback ladder as ChatPersistence: never
            // brick launch over analytics data.
            let aside = url.appendingPathExtension("corrupt-\(Int(Date().timeIntervalSince1970))")
            try? FileManager.default.moveItem(at: url, to: aside)
            if let retried = try? ModelContainer(for: schema, configurations: config) {
                return retried
            }
            let memory = ModelConfiguration(schema: schema, isStoredInMemoryOnly: true)
            return try! ModelContainer(for: schema, configurations: memory)
        }
    }()

    private static let queue = DispatchQueue(label: "Cuate.SpendLedger", qos: .utility)

    /// Appends one record. Fire-and-forget from any thread.
    static func append(kind: SpendKind, provider: String, model: String,
                       usage: TokenUsage = TokenUsage(), units: Double = 0,
                       costUSD: Double?, isEstimate: Bool = false) {
        queue.async {
            let ctx = ModelContext(container)
            ctx.insert(SDSpendRecord(
                kindRaw: kind.rawValue, provider: provider, model: model,
                inputTokens: usage.inputTokens, outputTokens: usage.outputTokens,
                cacheReadTokens: usage.cacheReadTokens, cacheWriteTokens: usage.cacheWriteTokens,
                reasoningTokens: usage.reasoningTokens,
                units: units, costUSD: costUSD, isEstimate: isEstimate
            ))
            try? ctx.save()
            Diagnostics.log("spend", "append kind=\(kind.rawValue) provider=\(provider) model=\(model) cost=\(costUSD.map { String(format: "%.6f", $0) } ?? "nil") est=\(isEstimate)")
        }
    }

    /// Fetches records in [from, to) as value types; completion runs on the
    /// ledger queue — callers hop to main themselves.
    ///
    /// Hermes agent records are excluded here, at the single point every
    /// Costs aggregate reads through: the gateway pays for its own model
    /// calls, and its run-cumulative usage frames (every tool-loop call
    /// re-counts the whole prompt) drowned the token stats. New agent turns
    /// are no longer recorded at all; this filter hides the ones already in
    /// the ledger.
    static func fetch(from: Date, to: Date, completion: @escaping ([SpendRecordValue]) -> Void) {
        queue.async {
            let ctx = ModelContext(container)
            let hermes = ProviderID.hermes.rawValue
            let predicate = #Predicate<SDSpendRecord> {
                $0.timestamp >= from && $0.timestamp < to && $0.provider != hermes
            }
            let descriptor = FetchDescriptor<SDSpendRecord>(
                predicate: predicate,
                sortBy: [SortDescriptor(\.timestamp)]
            )
            let rows = (try? ctx.fetch(descriptor)) ?? []
            completion(rows.map { row in
                SpendRecordValue(
                    id: row.id, timestamp: row.timestamp,
                    kind: SpendKind(rawValue: row.kindRaw) ?? .chat,
                    provider: row.provider, model: row.model,
                    usage: TokenUsage(
                        inputTokens: row.inputTokens, outputTokens: row.outputTokens,
                        cacheReadTokens: row.cacheReadTokens, cacheWriteTokens: row.cacheWriteTokens,
                        reasoningTokens: row.reasoningTokens
                    ),
                    units: row.units, costUSD: row.costUSD, isEstimate: row.isEstimate
                )
            })
        }
    }

    /// Timestamp of the oldest record (bounds the month picker), nil if empty.
    /// Same Hermes exclusion as `fetch` — an all-agent month would otherwise
    /// extend the picker into months that render empty.
    static func earliestTimestamp(completion: @escaping (Date?) -> Void) {
        queue.async {
            let ctx = ModelContext(container)
            let hermes = ProviderID.hermes.rawValue
            var descriptor = FetchDescriptor<SDSpendRecord>(
                predicate: #Predicate { $0.provider != hermes },
                sortBy: [SortDescriptor(\.timestamp)]
            )
            descriptor.fetchLimit = 1
            completion((try? ctx.fetch(descriptor))?.first?.timestamp)
        }
    }
}

// MARK: - UI-facing store

/// Main-actor aggregates for the Costs tab + the soft monthly budget. The
/// heavy lifting (fetch) happens on the ledger queue; this object only holds
/// published snapshots. Follows the ImageAddonSettings singleton pattern.
@MainActor
final class SpendStore: ObservableObject {
    static let shared = SpendStore()
    private let defaults = UserDefaults.standard

    /// Average tokens per chat message (current calendar month).
    struct AvgStats {
        let input: Int
        let output: Int
        let count: Int
    }

    /// Spend since app launch (all kinds), not persisted.
    @Published private(set) var sessionUSD: Double = 0
    /// Current-calendar-month totals (kept fresh by record()/refresh()).
    @Published private(set) var currentMonthUSD: Double = 0
    @Published private(set) var todayUSD: Double = 0
    /// Ø input/output tokens per chat message this month (nil until data).
    @Published private(set) var currentMonthAvg: AvgStats?
    /// Records of the month the Costs tab is looking at.
    @Published var selectedMonth: Date
    @Published private(set) var selectedMonthRecords: [SpendRecordValue] = []
    @Published private(set) var earliestRecord: Date?

    /// Soft monthly budget in USD; 0 = off. Warns, never blocks.
    @Published var monthlyBudgetUSD: Double {
        didSet { defaults.set(monthlyBudgetUSD, forKey: "spend.monthlyBudgetUSD") }
    }

    private init() {
        monthlyBudgetUSD = defaults.double(forKey: "spend.monthlyBudgetUSD")
        selectedMonth = Calendar.current.dateInterval(of: .month, for: Date())?.start ?? Date()
        refresh()
    }

    // MARK: Recording façade

    /// Central entry point every spend site calls: appends to the ledger,
    /// updates the live counters, kicks the weekly price refresh, and returns
    /// the budget-warning text to surface (nil when no threshold was crossed).
    @discardableResult
    func record(kind: SpendKind, provider: String, model: String,
                usage: TokenUsage = TokenUsage(), units: Double = 0,
                costUSD: Double?, isEstimate: Bool = false) -> String? {
        SpendLedger.append(kind: kind, provider: provider, model: model,
                           usage: usage, units: units, costUSD: costUSD, isEstimate: isEstimate)
        PricingCatalog.refreshIfStale()

        let cost = costUSD ?? 0
        sessionUSD += cost
        let before = currentMonthUSD
        currentMonthUSD += cost
        todayUSD += cost
        return budgetWarning(before: before, after: currentMonthUSD)
    }

    /// One warning per threshold per month: 80% and 100% of the budget.
    private func budgetWarning(before: Double, after: Double) -> String? {
        guard monthlyBudgetUSD > 0 else { return nil }
        let monthKey = Self.monthKey(Date())
        for (fraction, keySuffix, textKey) in [(1.0, "100", "costs.budgetHit"), (0.8, "80", "costs.budgetNear")] {
            let threshold = monthlyBudgetUSD * fraction
            guard after >= threshold, before < threshold else { continue }
            let flag = "spend.warned\(keySuffix).\(monthKey)"
            guard !defaults.bool(forKey: flag) else { return nil }
            defaults.set(true, forKey: flag)
            return String(format: L(textKey), Self.usd(after), Self.usd(monthlyBudgetUSD))
        }
        return nil
    }

    // MARK: Aggregation

    /// Reloads the live counters and the selected month's records.
    func refresh() {
        let calendar = Calendar.current
        let now = Date()
        guard let month = calendar.dateInterval(of: .month, for: now),
              let day = calendar.dateInterval(of: .day, for: now) else { return }

        SpendLedger.fetch(from: month.start, to: month.end) { records in
            let total = records.reduce(0) { $0 + ($1.costUSD ?? 0) }
            let today = records.filter { $0.timestamp >= day.start }
                .reduce(0) { $0 + ($1.costUSD ?? 0) }
            // Ø tokens per chat message, current month (summary row scope —
            // deliberately NOT the picker-selected month).
            let chats = records.filter { $0.kind == .chat }
            let avg: AvgStats? = chats.isEmpty ? nil : AvgStats(
                input: chats.reduce(0) { $0 + $1.usage.inputTokens + $1.usage.cacheReadTokens + $1.usage.cacheWriteTokens } / chats.count,
                output: chats.reduce(0) { $0 + $1.usage.outputTokens } / chats.count,
                count: chats.count
            )
            DispatchQueue.main.async {
                self.currentMonthUSD = total
                self.todayUSD = today
                self.currentMonthAvg = avg
            }
        }
        SpendLedger.earliestTimestamp { earliest in
            DispatchQueue.main.async { self.earliestRecord = earliest }
        }
        reloadSelectedMonth()
    }

    func reloadSelectedMonth() {
        let calendar = Calendar.current
        guard let interval = calendar.dateInterval(of: .month, for: selectedMonth) else { return }
        SpendLedger.fetch(from: interval.start, to: interval.end) { records in
            DispatchQueue.main.async { self.selectedMonthRecords = records }
        }
    }

    func stepMonth(by delta: Int) {
        let calendar = Calendar.current
        guard let next = calendar.date(byAdding: .month, value: delta, to: selectedMonth) else { return }
        // Clamp between the first recorded month and the current month.
        let lower = earliestRecord.flatMap { calendar.dateInterval(of: .month, for: $0)?.start }
            ?? selectedMonth
        let upper = calendar.dateInterval(of: .month, for: Date())?.start ?? selectedMonth
        selectedMonth = min(max(next, lower), upper)
        reloadSelectedMonth()
    }

    // MARK: Formatting helpers

    nonisolated static func usd(_ value: Double) -> String {
        value < 0.01 && value > 0
            ? String(format: "$%.4f", value)
            : String(format: "$%.2f", value)
    }

    nonisolated static func monthKey(_ date: Date) -> String {
        let comps = Calendar.current.dateComponents([.year, .month], from: date)
        return "\(comps.year ?? 0)-\(comps.month ?? 0)"
    }
}
